### ISSUE TITLE
Support variables in transformations

### ISSUE THREAD

**Issue by gilkgit** [grafana/grafana#25469]
In my scenario I have a dashboard with a repeated chart panel.
The repeated panel queries the datasource for each repetition.
I'd like to query once and then use that main query in all repetitions by filtering to show only lines relevant to that iteration.
This could potentially be achieved if I could use transformations with the repeated variable.

**Comment by Bio2hazard** [comment]
Yes please!

I'm trying to use the `Filter by name` transform to only display relevant data based on a templating variable but it's not working because the field does not seem to support interpolation. 

As long as I've been using grafana the seemingly arbitrary nature of where templating variables work and where they don't has been a constant source of frustration for me. 👎

**Comment by miry** [comment]
It would be usefull to have interpolation in transfomration. There is my case;
I have timeseries datapoints. The resolution window is dynamic. I would like to calculate operation per second with transformation. In this case similar for influxdb I use the variable $__interval or $__interval_ms.

**Comment by thiagogcm** [comment]
That's exactly what I tried to achieve but got stuck due this limitation.
Any hint on where to look in order to implement this feature? I may give it a try

**Comment by Mugglephilip** [comment]
Also need this feature, since we need to calculate QPS for different intervals

**Comment by jtommi** [comment]
I'd add that it should be possible to prevent a variable from triggering a new query.
I have my main data table and then I have summary panels that use the result from the main table panel.
If I want to filter the whole dashboard, I wouldn't need to reload the data, simply filter through the transform, no requerying needed

**Comment by bacsog** [comment]
+1

**Comment by briend** [comment]
This would be super useful, especially for `Filter Data by Values` transformation. I have several panels and one of them shows a table from a CSV file using a CSV Datasource plugin. I want to filter this table to only show the relevant data for the selected variable; i.e. `serverName`.  Without this, the CSV plugin itself needs to support filtering.

**Comment by franq77** [comment]
Variables in transformations would be really nice to have. I am having a drop down with currencies and would like to multiply all values with the selected currency exchange rate.

**Comment by PriyanshBordia** [comment]
Is there any tentative date for this feature to be live.?

**Comment by tomhollingworth** [comment]
What about something as the panel query runner executes to update or call something on each transformer to do the replacement? 

Playing around in a local branch https://github.com/tomhollingworth/grafana/commit/3a5a859e78c3853d9173dc74078412b4ae7ad264

`public/app/features/query/state/PanelQueryRunner.ts`
```diff
private getTransformationsStream = (withTransforms: boolean): MonoTypeOperatorFunction<PanelData> => {
    return (inputStream) =>
      inputStream.pipe(
        mergeMap((data) => {
          if (!withTransforms) {
            return of(data);
          }
          const transformations = this.dataConfigSource.getTransformations();
          if (!transformations || transformations.length === 0) {
            return of(data);
          }

+         const templateSrv = getTemplateSrv();
+
+         // Iterate over transforms and apply any template variable updates
+         const transforms = cloneDeep(transformations).map((transformation: DataTransformerConfig<any>) => {
+
+           // This example just updates the 'Add field from calculation` transformer
+           if (transformation.options.binary) {
+             if (transformation.options.binary.left) {
+               transformation.options.binary.left = templateSrv.replace(transformation.options.binary.left);
+             }
+             if (transformation.options.binary.right) {
+               transformation.options.binary.right = templateSrv.replace(transformation.options.binary.right);
+             }
+           }
+           return transformation;
+         });

+         return transformDataFrame(transforms, data.series).pipe(map((series) => ({ ...data, series })));
-         return transformDataFrame(transformations, data.series).pipe(map((series) => ({ ...data, series })));

          return transformDataFrame(transformations, data.series).pipe(map((series) => ({ ...data, series })));
        })
      );
  };
```

**Comment by syepes** [comment]
+1

**Comment by JaimeAvalosIDA** [comment]
+1

**Comment by EggTronic** [comment]
+1

**Comment by robkae** [comment]
+1

**Comment by silentZaika** [comment]
+1

**Comment by conet** [comment]
I was really eager to try out 8.5.0 waiting for this to be fixed, unfortunately it only fixes the repeated panel use case, it does not fix the case where there is a single panel that does a join based on a column produced by a template variable, I've been following this ever since #34606.

Because the template variable value instead of it's name appears in the transformation field selector one can't switch values and expect an update without refreshing the entire dashboard:
<img width="1378" alt="8 5 0" src="https://user-images.githubusercontent.com/11311763/165759244-01b8790c-8cb6-4afb-b667-76f5a23496c7.png">

With #35624 this works as expected and I can switch the variable value (field name) and get a different group by and join:  
<img width="1383" alt="with #35624" src="https://user-images.githubusercontent.com/11311763/165759251-04754309-1ea8-4064-a695-a7c3d1976e5f.png">

Frankly I don't consider this closed.

**Comment by krishnamurth** [comment]
Do we have support for variable name in the `filter by name` transformation yet ?? I am using Grafana v8.5.0 beta1 version, and i don't see the variable substitution happening. Any suggestions would really help.

![graf](https://user-images.githubusercontent.com/43027742/168423841-1c05ea0c-d885-4d34-86fc-2ac765b56f8c.jpg)

`Versions` is my variable name whose value i would want to be substituted for picking the fields. Let me know if this could be done ?

**Comment by krishnamurth** [comment]
Hi @torkelo : The variable interpolation in "Filter by name" still does not work.. Also this issue https://github.com/grafana/grafana/issues/42267 is marked as duplicated pointing to this issue.. Could you please suggest if this is in the radar to support ?? My use-case is mentioned in the previous comment...

**Comment by richardvanduijn** [comment]
This feature would be very very powerful in many situations. Upvoting it!! +1

**Comment by Alex389-r** [comment]
I am trying to use the variables on the "Filter data by values" transformation, but I cant make it work.

"Machine_Name" is my variable but doesnt seem to work:

![image](https://user-images.githubusercontent.com/109079901/178260721-6cea9d3b-9dcf-4aee-872a-5af24e3c8ef3.png)

I have tested that "Machine_Name" contains the exact string I need

**Comment by Kirchen99** [comment]
If I understand correctly, the Pull Request https://github.com/grafana/grafana/pull/44486 which "fixes" this issue has only added the feature to the transformer `Add field from calculation`.

And here is my PR for the transformer `Filter by Name`:
https://github.com/grafana/grafana/pull/51804

Maybe someone would like to add this feature to more transformers. But I think it is better to make it consistent for all transformers.

**Comment by JP-Besong** [comment]
@Alex389-r I'm trying to use variables in the same `Filter by values` transformation but I ran into the same issues you had. 

Did you find a solution for this? Or another way to use variables to filter your data?

**Comment by Alex389-r** [comment]
> @Alex389-r I'm trying to use variables in the same `Filter by values` transformation but I ran into the same issues you had.
> 
> Did you find a solution for this? Or another way to use variables to filter your data?

Im sorry but I did not find anny solution to this. I created a discussion (https://github.com/grafana/grafana/discussions/52219) talking about this and hoping variables will get supported in al transformations.

For the moment variables only work on "Add field from calculation" as Kirchen said

**Comment by shaunroselt** [comment]
This feels like a must-have feature. Variables should be supported everywhere.

I hope this is implemented in the near future.

**Comment by dcarrion87** [comment]
Same here for us

**Comment by jinnerbichler** [comment]
Still not working.

**Comment by Kirchen99** [comment]
Maybe we could track this issue for the current state of support variables in transformations:
 https://github.com/grafana/grafana/issues/46428

**Comment by BlackScreen** [comment]
Ok, stumbled upon this too. Why is this not supported by any of the transformations? Please add that!

**Comment by yulipumal** [comment]
Same here!

**Comment by SuperPauly** [comment]
How has this issue been going for over 3 years? Variables should be interpreted in all input. To me they are not Variables.

**Comment by shaunroselt** [comment]
> It seems to be planed in Grafana v10 #46428

Looks like this has been pushed to v11+. It's not planned for v10 anymore 😢

**Comment by conet** [comment]
This is crazy, grafana broke a perfectly fine feature in 8.x and still haven't fixed it, making users loose tons of hours debugging this.

**Comment by jmodrako** [comment]
Same here!

**Comment by oscarkilhed** [comment]
With https://github.com/grafana/grafana/pull/75372 we've added support for dashboard variables in transformations where we thought it would make sense. This feature is behind the feature flag `featureToggleAdminPage`.

We're planning on supporting other variables as well as we continue to work on this.

Unfortunately, the original feature request is not completely covered by this, as dashboard variables are scoped by the original query. Other panels based on the same query will have the variable value set to the same value as in the origin panel. So repeating panels based on a query from another panel will all receive the variable value as `{foo, bar ,baz}` instead of  `foo` for the first panel `baz` for the second panel etc.

**Comment by conet** [comment]
Please also add support for variables in joins, for example two series grouped by a variable `$filed2` 
<img width="477" alt="Screenshot 2023-10-05 at 17 21 49" src="https://github.com/grafana/grafana/assets/11311763/cf2fe599-a26e-4bdd-9491-8f865f66f1a7">

In 8.5.4 with [this patch](https://github.com/grafana/grafana/pull/35624/files) I can use the variable as a join column.
<img width="621" alt="Screenshot 2023-10-05 at 17 28 46" src="https://github.com/grafana/grafana/assets/11311763/d3edd3e2-d9a2-4afd-82d9-e0e665da5809">

In 10.1.2 I can't, because the variable is expanded to it's current value, thus breaking the join when the variable changes.
<img width="1355" alt="Screenshot 2023-10-05 at 17 22 17" src="https://github.com/grafana/grafana/assets/11311763/8e979a06-6b6a-413f-93c4-5ea7032dd063">

I had a hard time understanding why such a feature was removed, labeled unwanted behavior, when to me it seems like the most natural and useful behavior one could get from grafana. I can't upgrade until this will be possible in 10.x or someone provides a patch that offers the same behavior.

**Comment by bronzato** [comment]
+1
I really appreciate this feature, its very hard when using repeat option getting stuck by flat fields

**Comment by ugniusramanauskas** [comment]
+1

**Comment by nid06** [comment]
+1

**Comment by fferraro87** [comment]
+1

**Comment by bjrne** [comment]
+1

**Comment by Noahnc** [comment]
+1

**Comment by JustJoostNL** [comment]
+1

**Comment by patrickelectric** [comment]
+1

**Comment by hothero** [comment]
+1, I need to transform the custom variable value from dash to underline

**Comment by Zalasik01** [comment]
+1

**Comment by PhilippDehner** [comment]
Can be closed because is included in Grafana 11?

**Comment by undefinedhuman** [comment]
Any updates on this? (Grafana Cloud User this doesn't seem to work) - we are using a Transformation (Filter data by values), based on the GitLab Enterprise Plugin data source. 

We experienced the value being replaced once (sort of permanently) and it is removing the variable reference in the transformation. 

Although we also tested it without multi select (Repeating and non-Repeating), we believe this issue might be exactly what we experience: [https://github.com/grafana/grafana/issues/82393](https://github.com/grafana/grafana/issues/82393)
